import_export/src/file/export.rs

```diff
@@ -0,0 +1,240 @@
+    /// table. Overwrites existing files, if any, to ensure it has the
+    /// latest catalog information.
+    ///
+    /// 1. `<output_directory>/table.<partition_id>.json`: pbjson
+    /// encoded data about the table (minimal now)
+    ///
+    /// 2. `<output_directory>/partition.<partition_id>.json`: pbjson
+    /// encoded data for each partition
+    async fn export_table_metadata(
+        &mut self,
+        output_directory: &Path,
+        table_id: i64,
+    ) -> Result<()> {
+        // write table metadata
+        //
+        // (Note that since there is way to get table metadata via
+        // catalog API yet, make an empty object)
+        let table_json = "{}";
+        let filename = format!("table.{table_id}.json");
+        let file_path = output_directory.join(&filename);
+        write_string_to_file(table_json, &file_path).await?;
+
+        // write partition metadata for the table
+        let partitions = self
+            .catalog_client
+            .get_partitions_by_table_id(table_id)
+            .await?;
+
+        for partition in partitions {
+            let partition_id = partition.id;
+            if self.should_export(partition_id) {
+                let partition_json = serde_json::to_string_pretty(&partition)?;
+                let filename = format!("partition.{partition_id}.json");
+                let file_path = output_directory.join(&filename);
+                write_string_to_file(&partition_json, &file_path).await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Exports a remote ParquetFile to:
+    ///
+    /// 1. `<output_directory>/<uuid>.parquet`: The parquet bytes
+    ///
+    /// 2. `<output_directory>/<uuid>.parquet.json`: pbjson encoded `ParquetFile` metadata
+    async fn export_parquet_file(
+        &mut self,
+        output_directory: &Path,
+        index: usize,
+        num_parquet_files: usize,
+        parquet_file: &ParquetFile,
+    ) -> Result<()> {
+        let uuid = &parquet_file.object_store_id;
+        let partition_id = parquet_file.partition_id;
+        let file_size_bytes = parquet_file.file_size_bytes as u64;
+
+        let index = index + 1;
+
+        // copy out the metadata as pbjson encoded data always (to
+        // ensure we have the most up to date version)
+        {
+            let filename = format!("{uuid}.{partition_id}.parquet.json");
+            let file_path = output_directory.join(&filename);
+            let json = serde_json::to_string_pretty(&parquet_file)?;
+            write_string_to_file(&json, &file_path).await?;
+        }
+
+        let filename = format!("{uuid}.{partition_id}.parquet");
+        let file_path = output_directory.join(&filename);
+
+        if fs::metadata(&file_path)
+            .await
+            .map_or(false, |metadata| metadata.len() == file_size_bytes)
+        {
+            println!("skipping file {index} of {num_parquet_files} ({filename} already exists)");
```

Comment:
The file already exists _with the same file size_ (so presumably the same content)... is there a chance someone would see this message and rely on its assurance that it won't overwrite merely existing files, even if they _aren't_ the same content? Probably unlikely given the UUID in the filename, buuuut...

That is, I think this would be better as something like:

```suggestion
            println!("skipping file {index} of {num_parquet_files} ({filename} already exists with the expected file size)");
```

Reply:
in 6519a6377 👍